pkg/server/endpoints/authorized_entry_fetcher.go

```diff
@@ -0,0 +1,173 @@
+	cache                *authorizedentries.Cache
+	clk                  clock.Clock
+	log                  logrus.FieldLogger
+	ds                   datastore.DataStore
+	cacheReloadInterval  time.Duration
+	pruneEventsOlderThan time.Duration
+	lastEventID          uint
+}
+
+func NewAuthorizedEntryFetcherWithEventsBasedCache(ctx context.Context, log logrus.FieldLogger, clk clock.Clock, ds datastore.DataStore, cacheReloadInterval, pruneEventsOlderThan time.Duration) (*AuthorizedEntryFetcherWithEventsBasedCache, error) {
+	log.Info("Building in-memory entry cache")
+	cache, err := buildCache(ctx, ds)
+	if err != nil {
+		return nil, err
+	}
+	log.Info("Completed building in-memory entry cache")
+
+	return &AuthorizedEntryFetcherWithEventsBasedCache{
+		cache:                cache,
+		clk:                  clk,
+		log:                  log,
+		ds:                   ds,
+		cacheReloadInterval:  cacheReloadInterval,
+		pruneEventsOlderThan: pruneEventsOlderThan,
+	}, nil
+}
+
+func (a *AuthorizedEntryFetcherWithEventsBasedCache) FetchAuthorizedEntries(_ context.Context, agentID spiffeid.ID) ([]*types.Entry, error) {
+	return a.cache.GetAuthorizedEntries(agentID), nil
+}
+
+// RunUpdateCacheTask starts a ticker which rebuilds the in-memory entry cache.
+func (a *AuthorizedEntryFetcherWithEventsBasedCache) RunUpdateCacheTask(ctx context.Context) error {
+	for {
+		select {
+		case <-ctx.Done():
+			a.log.Debug("Stopping in-memory entry cache hydrator")
+			return nil
+		case <-a.clk.After(a.cacheReloadInterval):
+			err := a.updateCache(ctx)
+			if err != nil {
+				a.log.WithError(err).Error("Failed to update entry cache")
+			}
+		}
+	}
+}
+func (a *AuthorizedEntryFetcherWithEventsBasedCache) PruneEventsTask(ctx context.Context) error {
+	for {
+		select {
+		case <-ctx.Done():
+			a.log.Debug("Stopping event pruner")
+			return nil
+		case <-a.clk.After(a.pruneEventsOlderThan / 2):
+			a.log.Debug("Pruning events")
+			if err := a.pruneEvents(ctx, a.pruneEventsOlderThan); err != nil {
+				a.log.WithError(err).Error("Failed to prune events")
+			}
+		}
+	}
+}
+
+func (a *AuthorizedEntryFetcherWithEventsBasedCache) pruneEvents(ctx context.Context, olderThan time.Duration) error {
+	if err := a.ds.PruneRegistrationEntriesEvents(ctx, olderThan); err != nil {
+		return err
+	}
+	return a.ds.PruneAttestedNodesEvents(ctx, olderThan)
+}
+
+func (a *AuthorizedEntryFetcherWithEventsBasedCache) updateCache(ctx context.Context) error {
+	req := &datastore.ListRegistrationEntriesEventsRequest{
+		GreaterThanEventID: a.lastEventID,
+	}
+	resp, err := a.ds.ListRegistrationEntriesEvents(ctx, req)
+	if err != nil {
+		return err
+	}
+
+	for _, entryID := range resp.EntryIDs {
+		commonEntry, err := a.ds.FetchRegistrationEntry(ctx, entryID)
+		if err != nil {
+			return err
+		}
+
+		if commonEntry == nil {
+			a.cache.RemoveEntry(entryID)
+			a.lastEventID++
+			continue
+		}
+
+		entry, err := api.RegistrationEntryToProto(commonEntry)
```

Comment:
Hmm. I wonder if an error here should be logged and the entry skipped? Alternatively the entry could be removed from the cache...

Reply:
When does this routine return error?

Reply:
It would fail if there was some corrupt entry in the database. I'd prefer to skip that kind of entry instead of killing the whole event processing flow.

Reply:
Updated to remove entry and continue if this api returns error.